### ISSUE TITLE
Getting Battery Level Via Command Line

### ISSUE THREAD

**Issue by JCsplash** [jariz/Akku#59]
Hey @jariz  thanks for the great library! I’d just like to be the charge of my Bose QC35’s from mac’s terminal / command line.  

Is there a simple way to manually  put everything i need into one file that i can run via command line?

**Comment by jariz** [comment]
This is a good question! 
There's a [undocumented `--standalone` option](https://github.com/jariz/Akku/blob/master/AkkuHelper/Helper/Helper.swift#L36-L44) used for debugging the helper app that could be used for this, which can be ran with:
```
sudo /Applications/Akku.app/Contents/Library/LaunchServices/io.jari.AkkuHelper --standalone
```
This causes the helper to immediately start reading through the bluetooth history instead of waiting for the main Akku app to connect through XPC.

Example output looks something along the lines of:
```
Added connection, handle: 0C
Got connection request, ID: 18, handle: 0C
Added channel, CID: 01, sourceCID: 42, destinationCID: 606 handle: 0C
Got battery command: docked = nil, percentage = Optional(20)
```

Because this output is mostly used for debugging, it's not formatted in a very machine-readable way.
Do you have any suggestions on what kind of formatting you'd like? 
Or is this enough for your usecase?

**Comment by JCsplash** [comment]
All i need is the percentage, so this works fine. A few questions though. 

1. How energy intensive would this operation be? 
2. How fast does the operation take?
3. Does it interfere with the currently playing audio?

I plan to display the battery percentage on my touch bar with BetterTouchTool. Do you think it’s sustainable to fetch battery level every 1 second or say every 10 seconds?

**Comment by jariz** [comment]
Energy impact should be minimal, 
Akku itself runs this exact same operation [every 5 seconds](https://github.com/jariz/Akku/blob/c7c25502877515d0e322c0d2d0901a6436756b51/AkkuHelper/Bluetooth/Driver.swift#L130) and it hardly has an impact on the system. At worst, it uses 1-4% CPU whilst running this operation in my experience.
I just timed it with the unix time utility, and it takes about 0.09s total.

Akku does not interfere with currently playing audio, therefore this won't either.
Do keep in mind that playing audio muddies up the buffer that Akku reads from, therefore running this command every 5 seconds and caching the most recent result is recommended; there's a good chance that within 10 secs of the headset reporting it's battery status, it's already out of the buffer because of the audio data replacing it.

You can force your headset to send it's battery status by simply reconnecting. Otherwise it will only do a report if the battery state changes.

**Comment by JCsplash** [comment]
Got it, thank you. Does this approach still require installing Akku + the Helper app? Or can i just copy a few files from the source code to make this work?

**Comment by jariz** [comment]
Taking the executable out of the `Akku.app` I mentioned should be enough. It can run on it's own without any additional files.
Keep in mind that it needs to be ran as root. (sudo)

**Comment by JCsplash** [comment]
Awesome thanks @jariz Works great and it's very fast as well! 

Do you know why the battery percentage line is repeated multiple times? Initially it was 1 line but as i ran it more times, i got extra lines.

Also to clarify, about your comment on the buffer - you're saying that if the buffer's out (due to music playing), the battery will not report unless there's a change in percentage right?

```
Added connection, handle: 0B
Got connection request, ID: 93, handle: 0B
Added channel, CID: 01, sourceCID: 42, destinationCID: 442 handle: 0B
Got battery command: docked = nil, percentage = Optional(80)
Got battery command: docked = nil, percentage = Optional(80)
Got battery command: docked = nil, percentage = Optional(80)
Got battery command: docked = nil, percentage = Optional(80)
```

**Comment by jariz** [comment]
The device _should_ only report the battery if:

- It connects
- The state changes

But honestly it can very much differ per hardware, if it decides to send it multiply times on a timer or something it can very much do so 😛

**Comment by JCsplash** [comment]
From my testing it seems like the battery is only reported once when it first connects, but not again - unless i disconnect and reconnect the speaker. 

Are you saying this behavior is normal for some devices? it seems like the battery changes are reported in real time on my iphone though

**Comment by jariz** [comment]
Errr. Well I realise now guess there is a problem with this approach.
During normal usage, the app keeps track of all connection that have been made over time, so it will know which device is reporting what state. 

However, in the way you're using the app right now, it only reads through the buffer and nothing more. 
Basically, It's starts off with a clean slate every time it starts, so if it runs into any battery commands, it won't log them because it won't know from which device it originates, because it did not keep track of the connection ID's because they can't immediately be found in the buffer.

TLDR: I could maybe add a flag that instead of dumping everything to the console and qutting, keeps running and logging new battery states to the console, but not sure if your touchbar app plays nice with that.

**Comment by JCsplash** [comment]
Well, I’m calling ``` sudo Applications/Akku.app/Contents/Library/LaunchServices/io.jari.AkkuHelper --standalone ``` every couple of seconds so the current approach should work right? If say after 5 seconds the battery goes down 1% and i run the same sudo command, I should get the updated battery right? So far im not, and i wonder if its because the helper didn't fully quit and it has cached data somewhere.

Also I’m a bit uneasy about having to use sudo every time. Can you explain why sudo is needed and if you’ve seen any approaches to get battery level without root / sudo?

**Comment by JCsplash** [comment]
Hey @jariz  after more testing it seems like a lot of devices simply don’t report battery level frequently, including BoseQC35s and UE Speakers. They report every 10% increase / decrease - not Akku’s fault. Thanks for all the help though and for the great library!
